### PR TITLE
Add NETTY_CONNECTION_SEND_REQUEST_LATENCY to broker timer metrics

### PR DESCRIPTION
**Pull request by jackjlli**
## Description
Currently even though there is a scatter & gather broker metric, it's hard to tell whether the time is mostly spent in Netty connection or spent in server query processing. 

This PR:
* adds `NETTY_CONNECTION_SEND_REQUEST_LATENCY` to broker timer, so that transport latency  from broker to server can be tracked.
* turns `NETTY_CONNECTION_SEND_RESPONSE_LATENCY` into a table-level server timer metric, so that latencies can be measured for different tables.
* since the `TOTAL_QUERY_TIME` is already emitted on server, thus there is no need to add an extra metric in broker.

This PR adds these two more metrics into the broker query phase metrics to clearly denote two different values.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java [hunk 1]
```diff
@@ -81,5 +86,7 @@ void receiveDataTable(DataTable dataTable, int responseSize, int deserialization
     _dataTable = dataTable;
     _responseSize = responseSize;
     _deserializationTimeMs = deserializationTimeMs;
+    _serverProcessingTimeMs =
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:89]: So the way this is set on the server, this is ServerQueryPhase.QUERY_PROCESSING_TIME which is not the TOTAL_QUERY_TIME so won't include the conversion of DataTable.toBytes(). To measure the true wall clock time on the server, we need totalQueryTime I believe

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java [hunk 2]
```diff
@@ -86,6 +86,7 @@ public DataTableImplV3(int numRows, DataSchema dataSchema, Map<String, Map<Integ
    * Construct empty data table. (Server side)
    */
   public DataTableImplV3() {
+    super();
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java:89]: Why this change?

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java [hunk 3]
```diff
@@ -30,6 +30,8 @@
   QUERY_EXECUTION,
   QUERY_ROUTING,
   SCATTER_GATHER,
+  SERVER_PROCESSING,
+  NETTY_CONNECTION,
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java:34]: This I believe is used to measure the time it took for the server response to travel over the wire, arrive at broker and get deserialized. We already emit this from the server but I don't think it is at the table level
See the following code in InstanceRequestHandler

`_serverMetrics.addTimedValue(ServerTimer.NETTY_CONNECTION_SEND_RESPONSE_LATENCY, sendResponseLatencyMs, TimeUnit.MILLISECONDS);`

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java [hunk 4]
```diff
@@ -30,6 +30,8 @@
   QUERY_EXECUTION,
   QUERY_ROUTING,
   SCATTER_GATHER,
+  SERVER_PROCESSING,
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java:33]: Why do we need to emit a server metric from broker ? Each server is already emitting its own totalQueryTime so I don't think we need this. Will lead to more confusion IMO.

Comment (jackjlli) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java:33]: The primary aim for this PR is to extract the time spent on netty from the time spent on scatter_gather. We should have a metric to actually measure the network time. Now that the broker query phase has all the phases from end to end, it'd be good to have the server total query time here instead of leaving it to the server metrics.

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java (hunk 3 above)
Comment (jackjlli) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java:34]: The metric you mentioned doesn't exclude the time spent in processing the query in server, which isn't the purpose of this PR.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/SingleConnectionBrokerRequestHandler.java [hunk 5]
```diff
@@ -102,8 +103,17 @@ protected BrokerResponseNative processBrokerRequest(long requestId, BrokerReques
         dataTableMap.put(entry.getKey(), dataTable);
         totalResponseSize += serverResponse.getResponseSize();
       }
+      long serverProcessingTimeMs = serverResponse.getServerProcessingTimeMs();
+      if (serverProcessingTimeMs > -1) {
```
Comment (mcvsubbu) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/SingleConnectionBrokerRequestHandler.java:107]: Do we need this check?

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java (hunk 4 above)
Comment (mcvsubbu) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java:33]: We have often seen the case when either server or broker goes through a gc just as the response is being sent by server or received by broker. In this case, gc time will show up as  the "network time" you are measuring? On the server side, it will be the gc time included as "server time", but on the broker, if it receives the bytes from server and then slips into gc (presumably to copy the bytes in, so it is likely to happen), then we will see the gc time as netty time, right?

What exactly is the problem faced that we could not get an answer for using the current set of metrics?

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java (hunk 3 above)
Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java:34]: It excludes the query processing time. See code in Instance Request Handler

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java (hunk 4 above)
Comment (jackjlli) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java]: We couldn't avoid gc from happening. Wherever the gc happens is the place we should account for the time. I've just refactored the code in the latest push, which makes it more clear. Basically the scatter & gather consists of 3 parts: 
* request sent from broker to server
* query being processed (totalQueryTime)
* response sent from server to broker

If gc happens on any of the steps above, it should be measured in just that step.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/SingleConnectionBrokerRequestHandler.java (hunk 5 above)
Comment (jackjlli) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/SingleConnectionBrokerRequestHandler.java]: Code refactored.

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java (hunk 3 above)
Comment (jackjlli) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerQueryPhase.java]: Refactored the code to put `NETTY_CONNECTION_SEND_REQUEST_LATENCY` to broker timer metrics.

File: pinot-core/src/main/java/org/apache/pinot/core/transport/ServerChannels.java [hunk 6]
```diff
@@ -144,7 +148,12 @@ synchronized void sendRequest(InstanceRequest instanceRequest)
             System.currentTimeMillis() - startTime);
       }
       byte[] requestBytes = _serializer.serialize(instanceRequest);
-      _channel.writeAndFlush(Unpooled.wrappedBuffer(requestBytes), _channel.voidPromise());
+      long sendRequestStartTimeMs = System.currentTimeMillis();
+      _channel.writeAndFlush(Unpooled.wrappedBuffer(requestBytes)).addListener(f -> {
+        _brokerMetrics.addTimedTableValue(rawTableName, BrokerTimer.NETTY_CONNECTION_SEND_REQUEST_LATENCY,
+            (System.currentTimeMillis() - sendRequestStartTimeMs), TimeUnit.MILLISECONDS);
+        asyncQueryResponse.markRequestSent(serverRoutingInstance);
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/transport/ServerChannels.java:155]: You probably don't need to do markRequestSent.

File: pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java [hunk 7]
```diff
@@ -50,6 +51,14 @@ public int getSubmitDelayMs() {
     }
   }
 
+  public int getRequestSentDelayMs() {
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:54]: This is not being called anywhere so it won't be used and that's why I suggested in previous comment that you probably don't need to `asyncQueryResponse.markRequestSent(serverRoutingInstance);`

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:54]: If we need this value to be logged in ServerStats like other stats are logged in BaseBrokerRequestHandler, then let's change the toString implementation of ServerResponse since that is what is called by `AsyncQueryResponse.getStats()`

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerTimer.java [hunk 8]
```diff
@@ -29,7 +29,9 @@
   // metric tracking the freshness lag for consuming segments
   FRESHNESS_LAG_MS("freshnessLagMs", false),
 
-  NETTY_CONNECTION_SEND_RESPONSE_LATENCY("nettyConnection", true),
+  // The latency of sending the response from server to broker
+  NETTY_CONNECTION_SEND_RESPONSE_LATENCY("nettyConnection", false),
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerTimer.java:33]: I hope this is backward compatible change since name is not getting changed ?

Comment (jackjlli) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerTimer.java:33]: Yes, in fact the boolean isn't get used anywhere. So it's backward compatible.

File: pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java (hunk 7 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:54]: I think it's still worthwhile to keep it in ServerResponse, which makes it clearer about the delay from broker to server in the log. I'll add it to the toString method in the next push.

File: pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java [hunk 9]
```diff
@@ -69,13 +78,18 @@ public int getDeserializationTimeMs() {
   @Override
   public String toString() {
     return String
-        .format("%d,%d,%d,%d", getSubmitDelayMs(), getResponseDelayMs(), getResponseSize(), getDeserializationTimeMs());
+        .format("%d,%d,%d,%d,%d", getSubmitDelayMs(), getRequestSentDelayMs(), getResponseDelayMs(), getResponseSize(),
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:81]: better to add new metric at the end

File: pinot-core/src/main/java/org/apache/pinot/core/transport/AsyncQueryResponse.java [hunk 10]
```diff
@@ -73,7 +73,7 @@ public AsyncQueryResponse(QueryRouter queryRouter, long requestId, Set<ServerRou
    */
   public String getStats() {
     StringBuilder stringBuilder =
-        new StringBuilder("(Server=SubmitDelayMs,ResponseDelayMs,ResponseSize,DeserializationTimeMs)");
+        new StringBuilder("(Server=SubmitDelayMs,RequestSentDelayMs,ResponseDelayMs,ResponseSize,DeserializationTimeMs)");
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/transport/AsyncQueryResponse.java:76]: Add at the end?

**Review by siddharthteotia — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java [hunk 11]
```diff
@@ -50,6 +51,14 @@ public int getSubmitDelayMs() {
     }
   }
 
+  public int getRequestSentDelayMs() {
+    if (_sendRequestTimeMs != 0) {
+      return (int) (_sendRequestTimeMs - _startTimeMs);
+    } else {
+      return -1;
+    }
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:59]: @jackjlli , this seems incorrect. this is for time it took to send request on transport so although you are emitting in the metrics correctly , here since the delta is computed from start time, it gets recorded incorrectly

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/transport/ServerResponse.java:59]: The following is correct for existing `getSubmitDelayMs()` API

```
public int getSubmitDelayMs() {                          
  if (_submitRequestTimeMs != 0) {                       
    return (int) (_submitRequestTimeMs - _startTimeMs);  
  } else {                                               
    return -1;                                           
  }                                                      
}  
```

But what you are measuring is delay it took on the wire `getRequestSentDelayMs()`, so you should no rely on _startTime. Simply change the API to record the time


```
void markRequestSent(ServerRoutingInstance serverRoutingInstance, long timeMs) {
    _responseMap.get(serverRoutingInstance).markRequestSent(timeMs);
  }
```

**Review by siddharthteotia — APPROVED** [review]
